### ISSUE TITLE
History recall of a certain width and content corrupts command line entry

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#620]
This happens in both emacs and vi. Reproducer:

1. Make sure your terminal window is exactly 132 columns wide. (Verify with `echo $COLUMNS`)
2. Paste this exact command line into the terminal and press return:
   `arch/*/bin/ksh -c 'got=$( typeset -T Subsh_t=(typeset -i x); Subsh_t -a v=( (x=1) (x=2) (x=3) ); typeset -p v ); command -V Subsh_t'`
3. Press arrow up to retrieve the line from history.
4. If in emacs, position the cursor to the beginning with Ctrl+A (in vi, this happens automatically). The cursor is now mispositioned to immediately before the command line's first character, instead of on it.
5. Type something (in vi mode, press `i` first). The cursor does not advance when you type, it stays mispositioned to before the beginning of the line.

**Comment by pghvlaans** [comment]
Is the command line supposed to do anything? I just got `command: Subsh_t: not found`, and then going to the history item failed to reproduce the error (Linux, urxvt and xfce4-terminal).

**Comment by pghvlaans** [comment]
Update: I couldn't get it to work on MacOS or OpenBSD either.

**Comment by McDutchie** [comment]
No, what it does or doesn't do is irrelevant. It was a test I did at some point or other, then I noticed the line editor breaking. I am experiencing the problem on macOS 12.6.3 and can reproduce it on Apple's Terminal.app, iTerm2, and xterm (under X11).

**Comment by McDutchie** [comment]
Hmm. The bug also seems to be dependent on PS1. Try reproducing it with the default `PS1='$ '` or with a multi-line PS1 with that as the last line; that makes it reproducible for me (on macOS, not Linux).

**Comment by McDutchie** [comment]
Here is a video of the misbehaviour, just to prove I'm not crazy :)

https://user-images.githubusercontent.com/7469524/229332498-496039cc-b3bc-48fa-ba37-2124c5783b21.mp4

**Comment by pghvlaans** [comment]
OK, that did the trick. Thanks! This doesn't work on Linux or OpenBSD, but it does on both MacOS and Haiku. It appears that `PS1` just needs to be two characters or less:

* zero characters: Cursor goes to the last position prior to the command line; text enters in place
* one character: Cursor goes to the first character; entering text moves the previously-entered command line by one, and then text enters in place
* two characters: Cursor goes to the position prior to the first character; text enters in place (replacing the second character of `PS1`)

**Comment by pghvlaans** [comment]
FreeBSD has a different error. After pasting in the command and pressing return, navigating to the history line leaves the cursor after the command line in `vi` mode. In `emacs` mode, Ctrl+a has no effect. 

Entering any character (in `vi` mode, from insert) enters the character and then immediately repeats the command line. Navigating past the command line in history places the next command immediately after the command line, and leaves the cursor on the next line, offset by the length of `PS1` (which can apparently be any number of characters). Finally, moving the cursor to the right reveals the command line character by character.

**Comment by pghvlaans** [comment]
The bug is not reproducible when `multiline` is set to `off`. That should help narrow things down a bit.

**Comment by pghvlaans** [comment]
It appears that the trigger for this bug is `_` at or just before the edge of the window while in `multiline` for macOS and Haiku, whether pasted in or typed normally. Although that probably sounds very strange, in the function `ed_putchar`, at `edit.c:1025`:

    	if (buf[0] == '_' && size==1)
    	{
    		*dp++ = ' ';
    		*dp++ = '\b';
    	}

When this code is commented out (or put in `#else` from `#if SHOPT_MULTIBYTE`), the bug goes away. Does anyone know what the purpose of this was? Some kind of workaround for non-multibyte systems, maybe?

As a side note, this does not fix the issue on FreeBSD. Its bug is different (but very possibly related). There, it doesn't seem to matter what is typed in: corruption will occur if text hits the edge of the window in `multiline`.

EDIT: Also tested on Haiku.

EDIT 2: I'm going to do some more FreeBSD testing and open a separate issue.
